### DISCUSSION TITLE
Machinescope does not work for some applications

### DISCUSSION THREAD

**Discussion by fboesmans** [Romanitho/Winget-Install#86] [Q&A] [answered]
Hello

We are using this script to push software via Intune. For most of the applications, this is working great. But I have now stumbled on a problem where I'm trying to install Visual Studio Code machine wide.

So, in Intune I've created the package for VSCode with the following settings:
![image](https://github.com/Romanitho/Winget-Install/assets/104124972/f4252125-a704-4808-8e58-4ef806c8be6c)

For the detection rule, I also use the script that is provided in this repository. When the installation is done, the Company Portal is reporting that the application is installed correctly. But when I look for VSCode, I can't find it anywhere.

When I open an Powershell window in System context using PSExec, I can see that VSCode is installed in the 'User' scope.

I also checked the user settings file (settings.json) in the system context and it does say that the scope should be 'Machine'.

When I add '--scope=machine' to the install string, then VSCode is installed correctly under the system context.

When I run this script with a local admin account, then VSCode is also installed correctly. So, it seems that the user settings in the JSON file are not applied in the system context when I try to install VSCode.

When I test this script via Intune for Google Chrome, this application is installed machine wide.

Can you help me to solve this problem? Thank you in advance.

**Comment by Romanitho** [comment]
strange. I would suggest to force the install with `--scope=machine` for this app.

**Reply by dlundahl91** [reply]
Hi @Romanitho 

I am having the same troubles, with the exact same package, MS Visual Studio Code.

Where do I place this --scope=machine in order for this to work?
powershell.exe -NoProfile -ExecutionPolicy Bypass -File winget-install.ps1 -AppIDs Microsoft.VisualStudioCode --scope=machine (here for example?)

Thanks in advance

/Dan

**Reply by fboesmans** [reply]
Hi @dlundahl91 

You have to add it in the "Install-App" function on line 259:

        $WingetArgs = "install --id $AppID -e --accept-package-agreements --accept-source-agreements --scope=machine -h $AppArgs" -split " "

@Romanitho, I have now created 10 packages using your script and I have this issue with 2 applications, VSCode and Brave. 
I don't want to change the install string for each package because I have automated the creation of packages using your script. Hopefully you will be able to find a fix for it.

Many thanks for sharing your awesome script.

**Reply by Romanitho** [reply]
my bad, it is `--scope machine`

**Comment by fboesmans** [comment]
@Romanitho 

I am still experiencing the same problem as described in this discussion. But today I discovered something that makes me wonder if this is normal behavior.

Nothing has been changed to the program as shown in the screenshot. Today I note that in the logs the following part is skipped from your script:

![image](https://github.com/Romanitho/Winget-Install/assets/104124972/b4527f09-fde5-4a3a-add1-dc8ff0c664f9)

These steps do get completed when I change the install behavior to user.

When I open a powershell window via PSExec in system context and manually run the script, these steps are also completed correctly. But when the script is pushed via Intune, these are skipped.

Is this normal behavior?

**Reply by fboesmans** [reply]
Apparently this was fixed by removing the -NoProfile parameter. Unfortunately, though, this still does not solve my problem with the installation of VSCode.

**Reply by Romanitho** [reply]
I don't understand how the red part you've posted could be 'skipped' ?!

**Reply by fboesmans** [reply]
Hi @Romanitho 

I deploy your script via Intune. As you describe in the readme, I used the following as the install cmd:
`"%systemroot%\sysnative\WindowsPowerShell\v1.0\powershell.exe" -NoProfile -ExecutionPolicy Bypass -File winget-install.ps1 -AppIDs Notepad++.Notepad++`

Because of the -NoProfile parameter, the part in the red box was skipped:
![image](https://github.com/Romanitho/Winget-Install/assets/104124972/6f4c6c86-e778-4449-9c07-5a6fa944d379)

Like you can see in this screenshot, there is no message saying "Running with(out) admin rights". After I removed the -NoProfile parameter, the logs do say "Running with admin rights": 
![image](https://github.com/Romanitho/Winget-Install/assets/104124972/1d6fdca6-0222-4fee-aa32-bcbc60c5a78f)

I was hoping that this would fix my issue with the installation of VSCode because the Add-ScopeMachine function was always skipped but unfortunately the problem is still there.

**Reply by Romanitho** [reply]
are you sure you are using same version?? Logs seem totally different.

**Reply by fboesmans** [reply]
Yes, I'm sure that I'm using the same version. The full log from the last screenshot:

![image](https://github.com/Romanitho/Winget-Install/assets/104124972/46084f40-e463-4da6-ae8f-7b16fc6f4586)

But the script is working now as expected for most Winget installs. Only Visual Studio Code and a few other applications aren't installing correctly in system context. For example, VSCode is installing in the system profile and not machine wide.

**Reply by Romanitho** [reply]
VSCode installs in the system profile because it uses the user installer. Probably because Add-ScopeMachine is not executed.
To be honest I don't know why NoProfile switch would change the behavior of the script.

**Reply by KnifMelti** [reply]
Strange...
...this is what I get running as **System** the normal way:

![image](https://github.com/Romanitho/Winget-Install/assets/102996177/13eac39c-699b-4693-adfa-ae164babacea)
The machine scope has correctly been written to `$Env:windir\System32\config\systemprofile\AppData\Local\Microsoft\WinGet\Settings\settings.json` (missed that in the picture...)
Notice that it's the user installation that is downloaded and installed (even an error dialog for something wrong with the installer in this version at the end of the installation).
That installation ends up only for **System** in it's own context.

When installed with `-NoProfile -ExecutionPolicy Bypass -File "C:\virtualtest\winget-install.ps1" -AppIDs "Microsoft.VisualStudioCode --scope machine"` (you don't have to edit the `winget-install.ps1`) as in [Use Winget native parameters](https://github.com/Romanitho/Winget-Install#use-winget-native-parameters) everything get installed in **machine** context as it should

**Reply by Romanitho** [reply]
it seams that the settings.json is here now
`$Env:windir\System32\config\systemprofile\AppData\Local\Microsoft\WinGet\Settings\defautState\settings.json`
So I need to fix this to the scope to take effect

**Reply by fboesmans** [reply]
About the NoProfile, I will do some further testing in the coming days. But we also implemented the Windows 11 Intune CIS benchmark, maybe this is the reason why I get different results with the NoProfile parameter.
I will keep you updated.

**Reply by Romanitho** [reply, marked as answer]
https://github.com/Romanitho/Winget-Install/releases/tag/v1.11.3

**Reply by fboesmans** [reply]
I can confirm that the change fixed this problem! 

Many thanks!

**Comment by fboesmans** [comment]
> VSCode installs in the system profile because it uses the user installer. Probably because Add-ScopeMachine is not executed. To be honest I don't know why NoProfile switch would change the behavior of the script.
a
